File: flutter/beagle/test/default/url_builder_test.dart
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA;
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at;
 *
 *     http://www.apache.org/licenses/LICENSE-2.0;
 *
 * Unless required by applicable law or agreed to in writing, software;
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and;
 * limitations under the License.
 */

import 'package:beagle/default/url_builder.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('Given an UrlBuilder with a simple base url', () {
    const baseURL = 'http://base.url';
    final urlBuilder = UrlBuilder(baseURL);
    group('When build method is called', () {
      test('Then it should prepend baseUrl if path is relative', () {
        const path = '/relativePath';
        expect(urlBuilder.build(path), '$baseURL$path');
      });

      test('Then it should not add baseUrl if path is not relative', () {
        const absolutePath = 'absolutePath';
        expect(urlBuilder.build(absolutePath), absolutePath);
      });

      test('Then it should handle multiple / as a relative path', () {
        const path = '//weirdPath';
        expect(urlBuilder.build(path), '$baseURL$path');
      });

      test('Then it should handle root path', () {
        const path = '/';
        expect(urlBuilder.build(path), '$baseURL$path');
      });

      test('Then it should return path if path is empty', () {
        const path = '';
        expect(urlBuilder.build(path), '');
      });
    });

    group('Given an UrlBuilder with a base url ending with /', () {
      const baseURL = 'http://base.ending.with.slash/';
      final urlBuilder = UrlBuilder(baseURL);
      group('When build method is called', () {
        test(
            'Then it should not duplicate / when both baseUrl and path have it',
            () {
          const path = '/relativePath';
          expect(urlBuilder.build(path),
              'http://base.ending.with.slash/relativePath');
        });
      });
    });

    group('Given an UrlBuilder with an empty baseURL', () {
      final urlBuilder = UrlBuilder('');

      group('When build method is called', () {
        test('Then it should handle relative path when baseURL is empty', () {
          const path = '/relativePath';
          expect(urlBuilder.build(path), path);
        });

        test('Then it should handle absolute path when baseURL is empty', () {
          const absolutePath = 'absolutePath';
          expect(urlBuilder.build(absolutePath), absolutePath);
        });

        test('Then it should handle empty path when baseURL is empty', () {
          const path = '';
          expect(urlBuilder.build(path), '');
        });
      });
    });

    group('Given an UrlBuilder', () {
      final urlBuilder = UrlBuilder('');

      group('When build method is called with an encoded url', () {
        test('Then it should not encode URL', () {
          const encoded =
              'https://www.guiaviagensbrasil.com/imagens/Imagem%20do%20mar%20calma%20e%20belo%20da%20Praia%20da%20Engenhoca-Itacare-Bahia-BA.jpg';
          expect(urlBuilder.build(encoded), encoded);
        });
      });

      group('When build method is called with an uncoded url', () {
        test('Then it should encode URL', () {
          const original =
              'https://www.guiaviagensbrasil.com/imagens/Imagem do mar calma e belo da Praia da Engenhoca-Itacare-Bahia-BA.jpg';
          const encoded =
              'https://www.guiaviagensbrasil.com/imagens/Imagem%20do%20mar%20calma%20e%20belo%20da%20Praia%20da%20Engenhoca-Itacare-Bahia-BA.jpg';
          expect(urlBuilder.build(original), encoded);
        });
      });

      group('When build method is called with a url with diacritics', () {
        test('Then it should encode URL with diacritics', () {
          const original =
              'https://www.guiaviagensbrasil.com/imagens/Imagem do mar calma e belo da Praia da Engenhoca-Itacaré-Bahia-BA.jpg';
          const encoded =
              'https://www.guiaviagensbrasil.com/imagens/Imagem%20do%20mar%20calma%20e%20belo%20da%20Praia%20da%20Engenhoca-Itacar%C3%A9-Bahia-BA.jpg';
          expect(urlBuilder.build(original), encoded);
        });
      });
    });
  });
}

File: android/processor/src/test/java/br/com/zup/beagle/android/compiler/beaglesdk/UrlBuilderTest.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.android.compiler.beaglesdk

import br.com.zup.beagle.android.compiler.BeagleSetupProcessor.Companion.BEAGLE_SETUP_GENERATED
import br.com.zup.beagle.android.compiler.extensions.compile
import br.com.zup.beagle.android.compiler.mocks.BEAGLE_CONFIG_IMPORTS
import br.com.zup.beagle.android.compiler.mocks.LIST_OF_URL_BUILDER
import br.com.zup.beagle.android.compiler.mocks.SIMPLE_BEAGLE_CONFIG
import br.com.zup.beagle.android.compiler.mocks.VALID_URL_BUILDER
import br.com.zup.beagle.android.compiler.mocks.VALID_URL_BUILDER_BEAGLE_SDK
import br.com.zup.beagle.android.compiler.processor.BeagleAnnotationProcessor
import com.tschuchort.compiletesting.KotlinCompilation
import com.tschuchort.compiletesting.SourceFile
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Path

@DisplayName("Given Beagle Annotation Processor")
internal class UrlBuilderTest {

    @TempDir
    lateinit var tempPath: Path

    @DisplayName("When register url builder")
    @Nested
    inner class RegisterUrlBuilder {

        @Test
        @DisplayName("Then should add the url builder in beagle sdk")
        fun testGenerateUrlBuilderCorrect() {
            // GIVEN
            val kotlinSource = SourceFile.kotlin(
                FILE_NAME, BEAGLE_CONFIG_IMPORTS + VALID_URL_BUILDER + SIMPLE_BEAGLE_CONFIG)

            // WHEN
            val compilationResult = compile(kotlinSource, BeagleAnnotationProcessor(), tempPath)

            // THEN
            val file = compilationResult.generatedFiles.find { file ->
                file.name.startsWith(BEAGLE_SETUP_GENERATED)
            }!!

            val fileGeneratedInString = file.readText().replace(REGEX_REMOVE_SPACE, "")
            val fileExpectedInString = VALID_URL_BUILDER_BEAGLE_SDK
                .replace(REGEX_REMOVE_SPACE, "")

            assertEquals(fileExpectedInString, fileGeneratedInString)
            assertEquals(KotlinCompilation.ExitCode.OK, compilationResult.exitCode)
        }

    }


    @DisplayName("When register url builder")
    @Nested
    inner class InvalidUrlBuilder {

        @Test
        @DisplayName("Then should show error with duplicate url builder")
        fun testDuplicate() {
            // GIVEN
            val kotlinSource = SourceFile.kotlin(
                FILE_NAME, BEAGLE_CONFIG_IMPORTS + LIST_OF_URL_BUILDER + SIMPLE_BEAGLE_CONFIG)

            // WHEN
            val compilationResult = compile(kotlinSource, BeagleAnnotationProcessor(), tempPath)


            // THEN
            assertEquals(KotlinCompilation.ExitCode.COMPILATION_ERROR, compilationResult.exitCode)
            Assertions.assertTrue(compilationResult.messages.contains(MESSAGE_DUPLICATE_URL_BUILDER))
        }

    }

    companion object {
        private const val FILE_NAME = "File1.kt"
        private val REGEX_REMOVE_SPACE = "\\s".toRegex()
        private const val MESSAGE_DUPLICATE_URL_BUILDER = "UrlBuilder defined multiple times: \n" +
            "public final class UrlBuilderTestTwo implements " +
            "br.com.zup.beagle.android.networking.urlbuilder.UrlBuilder {\n" +
            "             ^\n" +
            "  br.com.test.beagle.UrlBuilderTestTwo\n" +
            "  br.com.test.beagle.UrlBuilderTest\n" +
            "  \n" +
            "  You must remove one implementation from the application."
    }

}